src/lib.rs

```diff
@@ -443,64 +447,74 @@ impl<T> Deque<T> {
-        // thread (the worker) is the only one that grows and shrinks it.
-        let buf = unsafe { self.inner.buffer.load(Relaxed, epoch::unprotected()) };
-
-        atomic::fence(SeqCst);
+        atomic::fence(Ordering::SeqCst);
 
         // Load the top.
-        let t = self.inner.top.load(Relaxed);
+        let t = self.inner.top.load(Ordering::Relaxed);
 
         // Compute the length after the bottom was decremented.
         let len = b.wrapping_sub(t);
 
-        if len < 0 {
+        if len <= 0 {
             // The deque is empty. Restore the bottom back to the original value.
-            self.inner.bottom.store(b.wrapping_add(1), Relaxed);
-            None
-        } else {
-            // Read the value to be popped.
-            let mut value = unsafe { Some(buf.deref().read(b)) };
-
-            // Are we popping the last element from the deque?
-            if len == 0 {
-                // Try incrementing the top.
-                if self.inner
-                    .top
-                    .compare_exchange(t, t.wrapping_add(1), SeqCst, Relaxed)
-                    .is_err()
-                {
-                    // Failed. We didn't pop anything.
-                    mem::forget(value.take());
-                }
+            self.inner.bottom.store(b, Ordering::Relaxed);
+            return None;
+        }
 
-                // Restore the bottom back to the original value.
-                self.inner.bottom.store(b.wrapping_add(1), Relaxed);
-            } else {
-                // Shrink the buffer if `len` is less than one fourth of `self.inner.min_cap`.
-                unsafe {
-                    let cap = buf.deref().cap;
-                    if cap > self.inner.min_cap && len < cap as isize / 4 {
-                        self.inner.resize(cap / 2);
-                    }
-                }
+        // Load the buffer. The buffer doesn't have to be epoch-protected because the current
+        // thread (the worker) is the only one that grows and shrinks it.
+        let buf = unsafe {
+            self.inner
+                .buffer
+                .load(Ordering::Relaxed, epoch::unprotected())
+        };
+
+        // Read the value to be popped.
+        let mut value = unsafe { Some(buf.deref().read(b.wrapping_sub(1))) };
+
+        // Are we popping the last element from the deque?
+        if len == 1 {
+            // Try incrementing the top.
+            //
+            // FIXME(jeehoonkang): in fact, `Release` for success and `Acquire` for failure is
+            // enough. But the C11 standard 7.17.7.4 paragraph 2 requires that the success ordering
+            // >= failure ordering: "The failure argument shall be no stronger than the success
+            // argument."
+            if self.inner
+                .top
+                .compare_exchange(t, t.wrapping_add(1), Ordering::AcqRel, Ordering::Acquire)
```

Comment:
You should change the orderings to `compare_exchange(t, t.wrapping_add(1), Ordering::Release, Ordering::Relaxed)` and add a `fence(Ordering::Acquire)` in the error path.

An acquire fence is generally slower than an acquire load (I know this is the case for ARM64 and PPC at least), however this is in the error path which is only hit in the rare case of a race.

Reply:
I agree.  I changed it.  Thanks!